src/LLVM_intrinsics.jl

```diff
@@ -700,4 +730,32 @@ for (f, neutral) in [(:fadd, "0.0"), (:fmul, "1.0")]
     end
 end
 
+# See: https://llvm.org/docs/LangRef.html#id839
+@generated function reduce_add(x::LVec{N, Bool}) where {N}
```

Comment:
I think all arithmetic operations so far preserve the type of the input vector. But, like Julia, we can add special cases for `Bool` and reductions. But it would be good to have a consistent idea about how that special casing is made. For example, why can we do `reduce(+, v)` on a vector of `Bool` but not `+v` for example?

Reply:
Maybe `+v` should work?  My thinking is that `Vec{N,T}` should work like `T` as much as possible so that we can increase the chance that passing `Vec` to user-defined function just works.

Reply:
I'm surprised that `+v` wouldn't work – it should.

Reply:
>  My thinking is that Vec{N,T} should work like T as much as possible so that we can increase the chance that passing Vec to user-defined function just works.

Okay, so then you think we should implement the whole promotion system for bools as well? 

```
julia> true + true
2

julia> sin(true)
0.8414709848078965
```

etc?

Reply:
Yeah, I think it makes sense, at least at the level of `Vec`.

I'm not sure if people use `sin(::Vec{N,Bool})` but I suppose it'd be handled by the promotion mechanism that would handle thing like `sqrt(::Vec{N,Int})` and `log10(::Vec{N,Int})` which are less crazy to support, I think.

Reply:
No, automatic promotions should not happen. The problem is that converting e.g. `Float32` to `Float64` is cheap, but converting `Vec{8,Float32}` to `Vec{8,Float64}` is quite expensive. Most likely, this changes even the number of registers required to hold a single vector.

With `Bool`, many CPU architectures have different representations of vectors of booleans, essentially offering `Bool32`, `Bool64`, etc., with rather expensive conversion operations between them. Thus conversions from `Bool` to other types should also not be automatic. I've contemplated offering types such as `Bool32` in the API, but haven't done so yet – my hope is that LLVM is clever enough to not make this necessary. (However, this doesn't handle how e.g. `Vector{Vec{8,Bool}}` should be efficiently stored; that's an unsolved problem.)

Reply:
> No, automatic promotions should not happen.

That effectively means that `Vec{N, Bool}` is not a computational type (which is the reason that e.g. `+` does not work on it right now). The sum of the ones in a `mask` is a useful thing to have though so special casing that (like in this PR) but not extending it to more operations might make sense.

Reply:
Agreed.